plutus-core/satint/src/Data/SatInt.hs

```diff
@@ -25,255 +25,61 @@ import Data.Csv
-                   !y' = y -# delta  -- x1 <= y' <= y; hence y' is representable
-
-                   -- Invariant: x <= y
-                   -- Note that: z <= y' => z + delta won't overflow
-                   -- so we are guaranteed not to overflow if/when we recurse
-                   go_up x | isTrue# (x ># y') = [SI (I# x)]
-                           | otherwise = SI (I# x) : go_up (x +# delta)
-               in SI (I# x1) : go_up x2
-
--- Requires x2 >= x1
-{-# INLINE [0] efdtIntUpFB #-}
-efdtIntUpFB :: (SatInt -> r -> r) -> r -> Int# -> Int# -> Int# -> r
-efdtIntUpFB c n x1 x2 y    -- Be careful about overflow!
- | isTrue# (y <# x2) = if isTrue# (y <# x1) then n else SI (I# x1) `c` n
- | otherwise = -- Common case: x1 <= x2 <= y
-               let !delta = x2 -# x1 -- >= 0
-                   !y' = y -# delta  -- x1 <= y' <= y; hence y' is representable
-
-                   -- Invariant: x <= y
-                   -- Note that: z <= y' => z + delta won't overflow
-                   -- so we are guaranteed not to overflow if/when we recurse
-                   go_up x | isTrue# (x ># y') = SI (I# x) `c` n
-                           | otherwise = SI (I# x) `c` go_up (x +# delta)
-               in SI (I# x1) `c` go_up x2
-
--- Requires x2 <= x1
-efdtIntDn :: Int# -> Int# -> Int# -> [SatInt]
-efdtIntDn x1 x2 y    -- Be careful about underflow!
- | isTrue# (y ># x2) = if isTrue# (y ># x1) then [] else [SI (I# x1)]
- | otherwise = -- Common case: x1 >= x2 >= y
-               let !delta = x2 -# x1 -- <= 0
-                   !y' = y -# delta  -- y <= y' <= x1; hence y' is representable
-
-                   -- Invariant: x >= y
-                   -- Note that: z >= y' => z + delta won't underflow
-                   -- so we are guaranteed not to underflow if/when we recurse
-                   go_dn x | isTrue# (x <# y') = [SI (I# x)]
-                           | otherwise = SI (I# x) : go_dn (x +# delta)
-   in SI (I# x1) : go_dn x2
-
-
--- Requires x2 <= x1
-{-# INLINE [0] efdtIntDnFB #-}
-efdtIntDnFB :: (SatInt -> r -> r) -> r -> Int# -> Int# -> Int# -> r
-efdtIntDnFB c n x1 x2 y    -- Be careful about underflow!
- | isTrue# (y ># x2) = if isTrue# (y ># x1) then n else SI (I# x1) `c` n
- | otherwise = -- Common case: x1 >= x2 >= y
-               let !delta = x2 -# x1 -- <= 0
-                   !y' = y -# delta  -- y <= y' <= x1; hence y' is representable
-
-                   -- Invariant: x >= y
-                   -- Note that: z >= y' => z + delta won't underflow
-                   -- so we are guaranteed not to underflow if/when we recurse
-                   go_dn x | isTrue# (x <# y') = SI (I# x) `c` n
-                           | otherwise = SI (I# x) `c` go_dn (x +# delta)
-               in SI (I# x1) `c` go_dn x2
-
--- The following code is copied/adapted from GHC.Real.
-
-instance Real SatInt where
-  toRational (SI x) = toInteger x % 1
-
-instance Integral SatInt where
```

Comment:
Not sure removing this is a good idea. We perhaps don't need, for example, integer division on `SatInt` at the moment, but if we do in the future, it's easy to overlook the possible overflow and use ``unSatInt x `div` unSatInt y``. The instance should also be included if we ever want to publish it as a separate library.

Reply:
I would've kept `Integral`, if only not its `Enum` constraint which brings in a ton of complexity, which I don't want to even think about.  I mean, we got `pred` and `succ` wrong, what guarantees do we have about the way more complicated stuff?

But maybe I should rename `toSatInt` to `unsafeToSatInt`? `unSatInt` is safe, it's only unsafe to embed an arbitrary, potentially underflown/overflown, `Int`.

Reply:
> But maybe I should rename toSatInt to unsafeToSatInt? unSatInt is safe, it's only unsafe to embed an arbitrary, potentially underflown/overflown, Int.

Great idea, I like it. Also `toSatInt` should have fewer usages than `unSatInt` so we don't litter the code with `unsafe`.

Reply:
I approve of getting rid of the division operations, since they let you take something saturated and convert it into something unsaturated without warning (ie, you could overflow and then convert back to a reasonable-looking result).